File: keetanetwork-anchor-client-wasi/bindings/csharp/KeetaNet.Anchor.Kyc/Sharable.cs
using System.Text.Json;

namespace KeetaNet.Anchor.Kyc.Crypto;

/// <summary>
/// A sealed, selectively disclosed subset of a KYC certificate's attributes.
/// The bundle and its derived state live inside the wasm core; this wrapper
/// holds only the handle and releases it on <see cref="Dispose"/>.
/// </summary>
public sealed class SharableCertificateAttributes : IDisposable
{
	private readonly WasmRuntime _runtime;
	private bool _disposed;

	/// <summary>The core-module handle backing this bundle.</summary>
	internal int Handle { get; }

	private SharableCertificateAttributes(WasmRuntime runtime, int handle)
	{
		_runtime = runtime;
		Handle = handle;
	}

	/// <summary>
	/// Prove or copy each attribute in <paramref name="names"/> from
	/// <paramref name="certificate"/> using the <paramref name="subject"/> account,
	/// bridging the trust chain with <paramref name="intermediates"/>, and seal the
	/// result. Grant a recipient before exporting.
	/// </summary>
	public static SharableCertificateAttributes FromCertificate(
		WasmRuntime runtime,
		KycCertificate certificate,
		Account subject,
		IEnumerable<Certificate>? intermediates = null,
		IEnumerable<string>? names = null)
	{
		int[] bridges = Handles(intermediates);
		string[] labels = (names ?? Enumerable.Empty<string>()).ToArray();
		return new(runtime, runtime.SharableFromCertificate(certificate.Handle, subject.Handle, bridges, labels));
	}

	/// <summary>
	/// Build like <see cref="FromCertificate"/>, additionally ingesting the
	/// caller-fetched external <paramref name="blobs"/> (raw fetched bytes keyed
	/// by reference id).
	/// </summary>
	public static SharableCertificateAttributes FromCertificateWithReferences(
		WasmRuntime runtime,
		KycCertificate certificate,
		Account subject,
		IReadOnlyDictionary<string, byte[]> blobs,
		IEnumerable<Certificate>? intermediates = null,
		IEnumerable<string>? names = null)
	{
		int[] bridges = Handles(intermediates);
		string[] labels = (names ?? Enumerable.Empty<string>()).ToArray();
		return new(runtime, runtime.SharableFromCertificateWithReferences(
			certificate.Handle, subject.Handle, bridges, labels, blobs));
	}

	/// <summary>Open a bundle from encoded container bytes, resolved with <paramref name="principals"/>.</summary>
	public static SharableCertificateAttributes FromEncoded(
		WasmRuntime runtime,
		byte[] data,
		IEnumerable<Account>? principals = null) =>
		new(runtime, runtime.SharableFromEncoded(data, AccountHandles(principals)));

	/// <summary>Open a bundle from a PEM envelope, resolved with <paramref name="principals"/>.</summary>
	public static SharableCertificateAttributes FromPem(
		WasmRuntime runtime,
		string pem,
		IEnumerable<Account>? principals = null) =>
		new(runtime, runtime.SharableFromPem(pem, AccountHandles(principals)));

	/// <summary>Grant <paramref name="accounts"/> access, invalidating the encoded form.</summary>
	public void GrantAccess(IEnumerable<Account> accounts) =>
		_runtime.SharableGrantAccess(Handle, AccountHandles(accounts));

	/// <summary>Revoke the account identified by its type-prefixed <paramref name="publicKey"/>.</summary>
	public void RevokeAccess(byte[] publicKey) => _runtime.SharableRevokeAccess(Handle, publicKey);

	/// <summary>The type-prefixed public keys of the accounts that can open the bundle.</summary>
	public IReadOnlyList<byte[]> Principals()
	{
		byte[] payload = _runtime.SharablePrincipals(Handle);
		int[][] raw = JsonSerializer.Deserialize<int[][]>(payload) ?? Array.Empty<int[]>();
		return raw.Select(values => Array.ConvertAll(values, value => (byte)value)).ToList();
	}

	/// <summary>The bundle's DER-encoded container bytes, requiring a granted recipient.</summary>
	public byte[] Export() => _runtime.SharableExport(Handle);

	/// <summary>The bundle exported as a PEM envelope.</summary>
	public string ToPem() => _runtime.SharableToPem(Handle);

	/// <summary>The embedded leaf certificate, as an independently owned object.</summary>
	public KycCertificate LeafCertificate() => KycCertificate.Adopt(_runtime, _runtime.SharableCertificate(Handle));

	/// <summary>The embedded intermediate certificate chain, as owned objects.</summary>
	public IReadOnlyList<Certificate> Intermediates()
	{
		byte[] payload = _runtime.SharableIntermediates(Handle);
		string[] pems = JsonSerializer.Deserialize<string[]>(payload) ?? Array.Empty<string>();
		return pems.Select(pem => Certificate.Parse(_runtime, pem)).ToList();
	}

	/// <summary>The names of the disclosed attributes.</summary>
	public IReadOnlyList<string> AttributeNames()
	{
		byte[] payload = _runtime.SharableAttributeNames(Handle);
		return JsonSerializer.Deserialize<string[]>(payload) ?? Array.Empty<string>();
	}

	/// <summary>The validated raw disclosed value for <paramref name="name"/>, or <c>null</c> when not disclosed.</summary>
	public byte[]? AttributeBuffer(string name)
	{
		byte[] value = _runtime.SharableAttributeBuffer(Handle, name);
		return value.Length == 0 ? null : value;
	}

	/// <summary>The schema-decoded semantic value for <paramref name="name"/>, or <c>null</c> when not disclosed.</summary>
	public byte[]? AttributeValue(string name)
	{
		byte[] value = _runtime.SharableAttributeValue(Handle, name);
		return value.Length == 0 ? null : value;
	}

	/// <summary>
	/// The inlined, digest-verified blob for reference <paramref name="id"/> on
	/// the disclosed attribute <paramref name="name"/>, or <c>null</c> when the
	/// attribute, entry, or matching reference node is absent.
	/// </summary>
	public byte[]? ReferenceBlob(string name, string id)
	{
		byte[] value = _runtime.SharableReferenceBlob(Handle, name, id);
		return value.Length == 0 ? null : value;
	}

	private static int[] Handles(IEnumerable<Certificate>? certificates) =>
		(certificates ?? Enumerable.Empty<Certificate>()).Select(certificate => certificate.Handle).ToArray();

	private static int[] AccountHandles(IEnumerable<Account>? accounts) =>
		(accounts ?? Enumerable.Empty<Account>()).Select(account => account.Handle).ToArray();

	public void Dispose()
	{
		if (_disposed)
		{
			return;
		}

		_disposed = true;
		_runtime.SharableFree(Handle);
	}
}

File: keetanetwork-anchor-client-wasi/src/p2/sharable.rs
//! The `sharable` selectively disclosed attributes resource of the P2 component.

use core::cell::RefCell;
use std::sync::Arc;

use keetanetwork_anchor::sharable_attributes::{
	ExternalBlobs, SharableCertificateAttributes as CoreSharableCertificateAttributes,
};
use keetanetwork_anchor_bindings::sharable_attributes as sharable_ops;

use super::account::AccountResource;
use super::certificate::CertificateResource;
use super::exports::keeta::anchor::certificates::KycCertificate as WitKycCertificate;
use super::exports::keeta::anchor::sharable::{
	Guest as SharableGuest, GuestSharableCertificateAttributes, KycCertificateBorrow,
	SharableCertificateAttributes as WitSharableCertificateAttributes,
};
use super::exports::keeta::client::crypto::{AccountBorrow, Certificate as WitCertificate, CertificateBorrow};
use super::kyc_certificate::KycCertificateResource;
use super::{collect_accounts, collect_certificates, CodedError, Component};

/// A sealed, selectively disclosed subset of a certificate's attributes.
pub(crate) struct SharableCertificateAttributesResource {
	bundle: RefCell<CoreSharableCertificateAttributes>,
}

impl SharableGuest for Component {
	type SharableCertificateAttributes = SharableCertificateAttributesResource;
}

impl GuestSharableCertificateAttributes for SharableCertificateAttributesResource {
	fn from_certificate(
		certificate: KycCertificateBorrow<'_>,
		subject: AccountBorrow<'_>,
		intermediates: Vec<CertificateBorrow<'_>>,
		names: Vec<String>,
	) -> Result<WitSharableCertificateAttributes, CodedError> {
		let certificate = &certificate.get::<KycCertificateResource>().certificate;
		let subject = Arc::clone(&subject.get::<AccountResource>().account);
		let intermediates = collect_certificates(&intermediates);
		let bundle = sharable_ops::from_certificate(certificate, &subject, &intermediates, &names)?;

		Ok(WitSharableCertificateAttributes::new(Self { bundle: RefCell::new(bundle) }))
	}

	fn from_certificate_with_references(
		certificate: KycCertificateBorrow<'_>,
		subject: AccountBorrow<'_>,
		intermediates: Vec<CertificateBorrow<'_>>,
		names: Vec<String>,
		blobs: Vec<(String, Vec<u8>)>,
	) -> Result<WitSharableCertificateAttributes, CodedError> {
		let certificate = &certificate.get::<KycCertificateResource>().certificate;
		let subject = Arc::clone(&subject.get::<AccountResource>().account);
		let intermediates = collect_certificates(&intermediates);
		let blobs = ExternalBlobs::from_iter(blobs);
		let bundle =
			sharable_ops::from_certificate_with_references(certificate, &subject, &intermediates, &names, blobs)?;

		Ok(WitSharableCertificateAttributes::new(Self { bundle: RefCell::new(bundle) }))
	}

	fn from_encoded(
		data: Vec<u8>,
		principals: Vec<AccountBorrow<'_>>,
	) -> Result<WitSharableCertificateAttributes, CodedError> {
		let principals = collect_accounts(&principals);
		let bundle = sharable_ops::from_encoded(&data, &principals)?;
		Ok(WitSharableCertificateAttributes::new(Self { bundle: RefCell::new(bundle) }))
	}

	fn from_pem(
		pem: String,
		principals: Vec<AccountBorrow<'_>>,
	) -> Result<WitSharableCertificateAttributes, CodedError> {
		let principals = collect_accounts(&principals);
		let bundle = sharable_ops::from_pem(&pem, &principals)?;
		Ok(WitSharableCertificateAttributes::new(Self { bundle: RefCell::new(bundle) }))
	}

	fn grant_access(&self, accounts: Vec<AccountBorrow<'_>>) -> Result<(), CodedError> {
		let accounts = collect_accounts(&accounts);
		sharable_ops::grant_access(&mut self.bundle.borrow_mut(), &accounts)?;
		Ok(())
	}

	fn revoke_access(&self, public_key: Vec<u8>) -> Result<(), CodedError> {
		sharable_ops::revoke_access(&mut self.bundle.borrow_mut(), &public_key)?;
		Ok(())
	}

	fn principals(&self) -> Result<Vec<Vec<u8>>, CodedError> {
		Ok(sharable_ops::principals(&self.bundle.borrow())?)
	}

	fn export_encoded(&self) -> Result<Vec<u8>, CodedError> {
		Ok(sharable_ops::export(&mut self.bundle.borrow_mut())?)
	}

	fn to_pem(&self) -> Result<String, CodedError> {
		Ok(sharable_ops::to_pem(&mut self.bundle.borrow_mut())?)
	}

	fn certificate(&self) -> Result<WitKycCertificate, CodedError> {
		let certificate = sharable_ops::certificate(&mut self.bundle.borrow_mut())?;
		Ok(WitKycCertificate::new(KycCertificateResource { certificate }))
	}

	fn intermediates(&self) -> Result<Vec<WitCertificate>, CodedError> {
		let intermediates = sharable_ops::intermediates(&mut self.bundle.borrow_mut())?;
		Ok(intermediates
			.into_iter()
			.map(|certificate| WitCertificate::new(CertificateResource { certificate }))
			.collect())
	}

	fn attribute_names(&self) -> Result<Vec<String>, CodedError> {
		Ok(sharable_ops::attribute_names(&mut self.bundle.borrow_mut())?)
	}

	fn attribute_buffer(&self, name: String) -> Result<Option<Vec<u8>>, CodedError> {
		Ok(sharable_ops::attribute_buffer(&mut self.bundle.borrow_mut(), &name)?)
	}

	fn attribute_value(&self, name: String) -> Result<Option<Vec<u8>>, CodedError> {
		Ok(sharable_ops::attribute_value(&mut self.bundle.borrow_mut(), &name)?)
	}

	fn reference_blob(&self, name: String, id: String) -> Result<Option<Vec<u8>>, CodedError> {
		Ok(sharable_ops::reference_blob(&mut self.bundle.borrow_mut(), &name, &id)?)
	}
}
